step2: 로또(자동)

sjjeong opened the pull request:
step1이 아직 완료되지 않았지만 step2와 영향도가 없어서 미리 진행 했습니다

sjjeong committed:
step2: 로또(자동)

kyucumber commented on src/main/kotlin/lotto/Lotto.kt:
```diff
@@ -0,0 +1,19 @@
+package lotto
+
+class Lotto(val numbers: List<Int>) {
```
`makeLotto` 메서드를 통해 만들면 항상 1-45까지의 6개의 숫자라는 제약이 지켜지겠지만, 생성자가 public으로 열려있어 임의로 생성할 수 있을 것 같아요.

1. `LottoNumber`를 추출해 1-45까지의 제약을 지키는 별도의 래핑된 클래스를 만들고 `Lotto` 클래스에 6개 및 중복 체크 validation을 추가한다.
2. 생성자를 private으로 변경하고 정적 팩터리인 makeLotto만을 이용해 생성하도록 한다.

위 두가지 방법 중 하나를 골라 제약 조건을 지킬 수 있도록 하면 좋을 것 같네요.

sjjeong replied:
LottoNumber 클래스 추가: 7f3baf8b685aa1842798a6b6f7c963050b277119
makeLotto 함수 제거: 5846c74d81602315067c0674db91e94494eaa8cc
Lotto 테스트 수정: ab7c325a04889dde2729567a65d7772262444186

kyucumber commented on src/main/kotlin/lotto/Prize.kt:
```diff
@@ -0,0 +1,39 @@
+package lotto
+
+sealed class Prize {
```
여기서 로또 당첨 결과에 대한 부분을 정의하는건 열거형(Enum)을 사용하는게 더 적절하지 않을까 하는 생각이 드네요.

다른 데이터를 보유하거나 다른 논리를 코드로 풀어낼 목적의 상수가 필요한 경우라면 sealed class가 적절하겠지만 지금은 money라는 데이터도 동일하고 별도로 다른 논리를 코드로 풀어낼 필요가 없어보여요.

정답은 없으니 이 부분에 대해 한번 고민해보시면 좋겠습니다.

sjjeong replied:
Prize를 enum으로 변경: 19a7ddbd7b62ef57958a581034228015d2e79fd3
Prize 테스트 코드: d22b1e6fd3d03d155126903d77bcf1e0d704e58b

kyucumber commented on src/main/kotlin/lotto/LottoResult.kt:
```diff
@@ -0,0 +1,34 @@
+package lotto
+
+class LottoResult {
+    private val prizeList = mutableListOf<Prize>()
```
모든 당첨 결과를 저장하고 있는건 다소 비효율적이지 않나 하는 생각이 드는데요.

정답이 정해진 것은 아니지만, groupingBy와 eachCount를 활용해 Map<Prize, Long> 형태로 자료구조를 바로 뽑아 활용하는건 어떨까요?

sjjeong replied:
LottoResult구현 변경: 680b841b4cc187f821d3e2a09aa76763f3fc8f97

kyucumber commented on src/main/kotlin/lotto/LottoVendingMachine.kt:
```diff
@@ -0,0 +1,16 @@
+    private const val LOTTO_PRICE = 1000
+
+    fun buyLotto(money: Int?): List<Lotto> {
+        requireNotNull(money) {
+            "구매 금액은 null이 아니어야 함"
+        }
+        require(money >= LOTTO_PRICE) {
+            "구매 금액은 1000 보다 커야 함"
+        }
```
여기서 의도한 대로 예외가 발생하는지 등에 대한 단위 테스트를 작성해보시면 좋겠습니다.

sjjeong replied:
LottoVendingMachine 관련 테스트 코드: c2869c573f82f907ed08ba3a874132d986575a0e, 0c52ba53c0c9837ef20f11e81919af2f8b92f811

kyucumber commented on src/main/kotlin/lotto/LottoResult.kt:
```diff
@@ -0,0 +1,34 @@
+package lotto
+
+class LottoResult {
```
이 부분도 단위 테스트를 작성해볼 수 있지 않을까요?

sjjeong replied:
LottoResult 테스트 코드 추가: 680b841b4cc187f821d3e2a09aa76763f3fc8f97

kyucumber commented on src/test/kotlin/lotto/LottoTest.kt:
```diff
@@ -0,0 +1,103 @@
+
+class LottoTest {
+
+    @Test
```
junit의 테스트 어노테이션을 사용하고 계시네요.

자동차 경주 미션 이후로 코틀린이 조금 익숙해지셨다면 [kotest의 여러 테스트 레이아웃](https://kotest.io/docs/framework/testing-styles.html) 및 [assertion](https://kotest.io/docs/assertions/assertions.html)을 활용해보셔도 좋을 것 같아요.

sjjeong replied:
assertion은 사용중인데 아직 레이아웃은 익숙하지가 않네요 ㅎㅎ
다음에 시도해보겠습니다~ 감사합니다!

kyucumber reviewed (changes requested):
안녕하세요

몇가지 코멘트 남겨두었으니 확인 부탁드립니다.

wotjd243 committed:
build: upgrade to kotlin 1.9.24 and gradle 8.7

sjjeong committed:
Merge branch 'main' into step2

sjjeong committed:
refactor: LottoNumber 추가

항상 1~45 이내의 숫자만 받을 수 있는 LottoNumber 추가

sjjeong committed:
refactor: makeLotto 함수 제거

makeLotto함수를 제거하고, init 블럭에서 lottoNumers의 유효성을 검사하도록 변경

sjjeong committed:
refactor: Lotto 테스트 변경

기존에는 Lotto 클래스에 대한 테스트가 아니라 LottoVendingMachiine에 대한 테스트로 검증했어서 변경함
다른 코드들은 추후에 이동할 예정

sjjeong committed:
refactor: Prize를 enum으로 변경

지금은 sealed class일 필요가 없어서 변경함

sjjeong committed:
refactor: PrizeTest 생성

LottoTest에 있던 Prize 관련 테스트 코드를 이동

sjjeong committed:
refactor: LottoResult 구현 변경

List가 아닌 Map 형태로 결과를 가지도록 변경

sjjeong committed:
refactor: LottoVendingMachineTest 파일 추가

LottoTest에 있던 LottoVendingMachine 관련 테스트들 이동

sjjeong committed:
test: LottoVendingMachineTest 에 경계테스트 추가

wisemuji commented on src/main/kotlin/lotto/Lotto.kt:
```diff
@@ -0,0 +1,26 @@
+        require(lottoNumbersSize == lottoNumbers.toSet().size)
+    }
+
+    constructor() : this((1..45).shuffled().take(6).sorted().map { LottoNumber(it) })
```
이건 정답이 없는 문제이지만, `LottoNumber`가 한 로또 어플리케이션 주기 동안 매번 인스턴스화되어야 할까요?

wisemuji commented on src/main/kotlin/lotto/Lotto.kt:
```diff
@@ -0,0 +1,26 @@
+    init {
+        val lottoNumbersSize = lottoNumbers.size
+        require(lottoNumbersSize == LOTTO_SIZE)
+        require(lottoNumbersSize == lottoNumbers.toSet().size)
```
`Lotto`의 `lottoNumbers` 자체를 `Set`으로 관리하는 것도 고려해보시면 어떨까요? 이 코멘트는 참고만 해주셔도 좋습니다.

wisemuji commented on src/main/kotlin/lotto/LottoNumber.kt:
```diff
@@ -0,0 +1,12 @@
+@JvmInline
+value class LottoNumber(val number: Int) {
+    init {
+        require(number in 1..45)
```
45라는 숫자를 여러 객체에서 들고 있는 것이 적절할까요? (저는 모든 중복이 나쁘다고 생각하진 않습니다)

wisemuji commented on src/main/kotlin/lotto/LottoNumber.kt:
```diff
@@ -0,0 +1,12 @@
+    }
+
+    override fun toString(): String {
+        return number.toString()
```
자동차 경주 5단계의 그림을 참고하여 도메인 로직에서의 뷰 의존성을 걷어내주세요!

![image](https://github.com/user-attachments/assets/384dd8a9-b829-4888-83a0-278f2fb8c8bb)

wisemuji commented on src/test/kotlin/lotto/LottoResultTest.kt:
다음 단계의 아래 요구 사항을 반영하시는 과정에서 테스트 코드가 수정될 수 있겠네요!
> 일급 컬렉션을 쓴다.

wisemuji reviewed (approved):
안녕하세요 석준님! 리뷰 반영을 잘 해주신 것 같아서 2단계는 이만 머지할게요 🙇 

조금 고민해보시면 좋을 내용을 코멘트로 달았으니 다음 단계 미션 진행해주시면서 같이 반영해주시면 됩니다!
